src/uu/expr/src/syntax_tree.rs

```diff
@@ -200,6 +194,38 @@ const PRECEDENCE: &[&[(&str, BinOp)]] = &[
     &[(":", BinOp::String(StringOp::Match))],
 ];
 
+#[derive(Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub enum NumOrStr {
+    Num(BigInt),
+    Str(String),
+}
+
+impl NumOrStr {
+    pub fn to_usize(self: NumOrStr) -> Option<usize> {
+        match self.to_bigint() {
+            Ok(num) => num.to_usize(),
+            Err(_) => None,
+        }
+    }
+
+    pub fn to_string(self: &NumOrStr) -> String {
```

Comment:
Does this need to take a reference? I think we can skip an allocation here if we just take ownership.